src/java/htsjdk/samtools/BAMIndexer.java

```diff
@@ -68,6 +72,10 @@ public BAMIndexer(final File output, final SAMFileHeader fileHeader) {
      */
     public BAMIndexer(final OutputStream output, final SAMFileHeader fileHeader) {
 
+        if (fileHeader.getSortOrder() != SAMFileHeader.SortOrder.coordinate) {
+            throw new SAMException("BAM file be coordinate-sorted for indexing.");
```

Comment:
Since these constructors are getting some logic in them I think it would be good to refactor them to call out to a private constructor that takes a BinaryBAMIndexWriter.